windows/src/developer/TIKE/compile/CompileKeymanWeb.pas

```diff
@@ -873,11 +890,31 @@ function TCompileKeymanWeb.JavaScript_Name(i: Integer; pwszName: PWideChar; UseN
       else Result := '_'; // Ensures we cannot overlap numbered instances
     while pwszName^ <> #0 do
     begin
-      if CharInSet(PChar(pwszName)^, SValidIdentifierCharSet)   // I3681
-        then Result := Result + PChar(pwszName)^
-        else Result := Result + '_';
+      if CharInSet(PChar(pwszName)^, SValidIdentifierCharSet) then  // I3681
+      begin
+        Result := Result + PChar(pwszName)^
+      end
+      else
+      begin
+        Result := Result + '_';
+        FChanged := True;
+      end;
       Inc(pwszName);
     end;
+    if not UseNameForRelease then
+    begin
+      // Ensure each transformed name is still unique
+      Result := Result + '_' + IntToStr(i);
+      if FChanged then
+        Result := Result + '/*'+string(p).Replace('*/', '*-/')+'*/'
+    end
```

Comment:
I may be missing something here.  We're emitting the original name when `not UseNameForRelease`?

I'd have thought the plan was that when `UseNameForRelease` is true but the option store's name isn't in the "valid identifier set", we'd use non-named JS variable names but inline the original name as a comment.

Reply:
Would avoid the need for a "breaking change" note, too.

Reply:
`UseNameForRelease` is perhaps a poor name for the parameter. When this flag is set, we need to emit the original name of the store, because it is being used as a string parameter for `KSAVE()`, `KLOAD()` -- so the store can persist across versions of the keyboard where the store indices may change. This is why I figured it safest to require "valid identifier characters" for these stores -- to reduce potential complications for saving them, e.g. character set woes in cookies, etc. It may be we can use the original name with appropriate escaping but I wonder if it is worth the effort?

Reply:
But this PR's description says:

> This implements @jahorton's suggestion of including the original name in a comment alongside the munged name, as well as appending the unique integer identifier of each symbol.

That suggestion was to _allow_ a partially-munged name (when necessary) to be used for `UseNameForRelease` (when true) by pairing it with an in-lined comment of the true name at all points of reference.  Then, the inlined comment can serve as said "appropriate escaping" while not causing store name overrides.

Reply:
Unless I am completely misunderstanding you, I don't think you are understanding what `UseNameForRelease` is doing. `UseNameForRelease` has nothing to do with debug mode or Javascript identifiers. It is set to true by the caller when we need a *string* representation of the store identifier to use for variable stores, to persist or load the value of the variable store.

So, this line of code:

```
+ 'a' > set(foo='1')
```

Gives us:

```
if(k.KKM(e, modCodes.VIRTUAL_KEY /* 0x4000 */, keyCodes.K_A /* 0x41 */)) {
      if(1){
        r=m=1;   // Line 23
        k.KDC(0,t);
        k.KSAVE("foo",this.s_foo_10);
      }
    }
```

See how the `k.KSAVE` function handles this with debug mode on. Now, with debug mode off:

```
if(k.KKM(e,16384,65)) {if(1){r=m=1;k.KDC(0,t);k.KSAVE("foo",this.s10);}}
```

The `"foo"` is preserved here because it's used as an identifier for persistence.

That's the only use of `UseNameForRelease`. The parameter name does makes sense, if you think of it as saying, use the store's name even when we are generating a release build (i.e. non-debug). But it's a bit confusing all the same.

Reply:
Ah, I see.   Yeah, the flag could _definitely_ use a better name, then.

Reply:
@mcdurdin is currently accepting suggestions for better names

Reply:
Since it's main use is for persistent storage / cookies, some ideas:

- `UsePersistentName`
- `AsPersistentID`
- `UseNameForCookie`
- `UseNameForPersistentStorage`

Reply:
`KeepNameForPersistentStorage` won for today.